src/sentry/incidents/subscription_processor.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import absolute_import
+
+import logging
+import operator
+from datetime import timedelta
+
+from django.conf import settings
+from django.db import transaction
+
+from sentry.incidents.logic import (
+    alert_aggregation_to_snuba,
+    create_incident,
+    update_incident_status,
+)
+from sentry.incidents.models import (
+    AlertRule,
+    AlertRuleAggregations,
+    AlertRuleThresholdType,
+    Incident,
+    IncidentStatus,
+    IncidentType,
+)
+from sentry.utils import metrics, redis
+from sentry.utils.dates import to_datetime
+
+
+logger = logging.getLogger(__name__)
+
+
+class SubscriptionProcessor(object):
+    """
+    Class for processing subscription updates for an alert rule. Accepts a subscription
+    and then can process one or more updates via `process_update`. Keeps track of how
+    close an alert rule is to alerting, creates an incident, and auto resolves the
+    incident if a resolve threshold is set and the threshold is triggered.
+    """
+
+    REDIS_TTL = int(timedelta(days=7).total_seconds())
+    # Each entry is a tuple in format (<alert_operator>, <resolve_operator>)
+    THRESHOLD_TYPE_OPERATORS = {
+        AlertRuleThresholdType.ABOVE: (operator.gt, operator.lt),
+        AlertRuleThresholdType.BELOW: (operator.lt, operator.gt),
+    }
+    ALERT_RULE_BASE_STAT_KEY = "{alert_rule:%s}:%%s"
+    ALERT_RULE_STAT_KEYS = ("last_update", "alert_triggered", "resolve_triggered")
+
+    def __init__(self, subscription):
+        self.subscription = subscription
+        try:
+            self.alert_rule = AlertRule.objects.get(query_subscription=subscription)
+        except AlertRule.DoesNotExist:
+            return
```

Comment:
Should this raise? The Processor is incomplete without a rule and is missing many of its attributes.

Reply:
We handle this later. If there's no `alert_rule` present, then we just skip trying to process the update. Ideally we'll also delete the subscription, since it's invalid, but I'll handle that as a follow up pr.